### PR TITLE
Better docs for Services and public IPs.

### LINKED ISSUE #2738: Explain/justify PublicIP field in service
It was added in #2319, but it's unclear what it's used for, who uses it, who sets it, etc.

### PR DESCRIPTION
**Pull request by thockin**
Fixes #2738

### CODE REVIEW

**Comment by thockin** [comment]
now merged.

File: pkg/api/v1beta3/types.go [hunk 1]
```diff
@@ -742,8 +742,9 @@ type ServiceSpec struct {
 
 	// CreateExternalLoadBalancer indicates whether a load balancer should be created for this service.
 	CreateExternalLoadBalancer bool `json:"createExternalLoadBalancer,omitempty"`
-	// PublicIPs are used by external load balancers.
-	PublicIPs []string `json:"publicIPs,omitempty"`
+	// PublicIPs are used by external load balancers, or can be set by
+	// users to handle external traffic that arrives at a node.
+	PublicIPs []string `json:"publicIPs,omitempty" description:"externally visible IPs (e.g. load balancers) that should be proxied to this service"`
```
Comment (bgrant0607) [inline pkg/api/v1beta3/types.go]: Heads-up @vishh

Comment (vishh) [inline pkg/api/v1beta3/types.go]: Ack.

File: docs/services.md [hunk 2]
```diff
@@ -2,64 +2,72 @@
...
-of `services` is to provide a bridge for non-Kubernetes-native applications to
-access backends without the need to write code that is specific to Kubernetes.
-A `service` offers clients an IP and port pair which, when accessed, redirects
-to the appropriate backends.  The set of `pods` targetted is determined by a label
-selector.
-
-As an example, consider an image-process backend which is running with 3 live
+Kubernetes [`Pods`](Pods.md) are mortal. They are born and they die, and they
+are not resurrected.  [`ReplicationControllers`](replication-controller.md) in
+particular create and destroy `Pods` dynamically.  While each `Pod` gets its own
```
Comment (bgrant0607) [inline docs/services.md]: Maybe add examples: node failures, auto-scaling, rolling updates.

File: docs/services.md [hunk 3]
```diff
@@ -2,64 +2,72 @@
...
-frontend client(s) do not need to know that.  The `service` abstraction
-enables this decoupling.
+they use.  While the actual `Pods` that comprise the backend set may change, the
+frontend clients should not need to manage that themselves.  The `Service`
+abstraction enables this decoupling.
 
-## Defining a service
+For Kubernetes-native applications, Kubernetes offers a simple `Endpoints` API
+that is updated whenever the set of `Pods` in a `Service` changes.  For
+non-native applications, Kubernetes offers a virtual-IP based bridge to Services
```
Comment (bgrant0607) [inline docs/services.md]: virtual-IP-based

File: docs/services.md [hunk 4]
```diff
@@ -109,24 +150,79 @@ REDIS_MASTER_PORT_6379_TCP_PORT=6379
 REDIS_MASTER_PORT_6379_TCP_ADDR=10.0.0.11
 ```
 
-This does imply an ordering requirement - any `service` that a `pod`
-wants to access must be created before the `pod` itself, or else the environment
-variables will not be populated.  This restriction will be removed once DNS for
-`services` is supported.
+This does imply an ordering requirement - any `Service` that a `Pod` wants to
```
Comment (bgrant0607) [inline docs/services.md]: Probably should bold this or something.

**Comment by bgrant0607** [comment]
Overall, LGTM. Just a few minor comments. It doesn't mention readiness probes, but I don't want to block this PR on that.

File: docs/services.md (hunk 2 above)
Comment (bgrant0607) [inline docs/services.md]: That is, examples of why pods would be created or destroyed.

Comment (thockin) [inline docs/services.md]: done in next push

File: docs/services.md (hunk 3 above)
Comment (thockin) [inline docs/services.md]: ok

File: docs/services.md (hunk 4 above)
Comment (thockin) [inline docs/services.md]: ok

**Comment by thockin** [comment]
new push is up with nits fixed

**Comment by bgrant0607** [comment]
LGTM

**Pull request merged**